src/Sarif/Query/ExpressionParser.cs

```diff
@@ -0,0 +1,354 @@
+                        if (text[terminatorIndex + 1] != terminator) { break; }
+
+                        // Copy to the StringBuilder without either quote
+                        if (unescapedForm == null) { unescapedForm = new StringBuilder(); }
+                        text.Substring(nextCopyFrom, terminatorIndex - nextCopyFrom).AppendTo(unescapedForm);
+
+                        // Include the second quote next time
+                        nextCopyFrom = terminatorIndex + 1;
+
+                        // Skip to look after the second quote next iteration
+                        terminatorIndex += 1;
+                    }
+                }
+
+                // Ensure the string is terminated
+                if (terminatorIndex == text.Length || text[terminatorIndex] != terminator) { throw new QueryParseException("string terminator", text); }
+
+                if (unescapedForm != null)
+                {
+                    if (nextCopyFrom < terminatorIndex) { text.Substring(nextCopyFrom, terminatorIndex - nextCopyFrom).AppendTo(unescapedForm); }
+                    text = text.Substring(terminatorIndex + 1);
+                    return unescapedForm.ToString();
+                }
+                else
+                {
+                    StringSlice resultSlice = text.Substring(1, terminatorIndex - 1);
+                    text = text.Substring(terminatorIndex + 1);
+                    return resultSlice;
+                }
+            }
+            else
+            {
+                // Read until whitespace or ')' (so closing a subexpression doesn't require a space)
+                int terminatorIndex = 0;
+                for (; terminatorIndex < text.Length; ++terminatorIndex)
+                {
+                    char c = text[terminatorIndex];
+                    if (c == ')' || IsWhitespace(c)) { break; }
+                }
+
+                // Consume and return the string
+                StringSlice result = text.Substring(0, terminatorIndex);
+                text = text.Substring(terminatorIndex);
+                return result;
+            }
+        }
+
+        private static Literal<ExpressionToken> StartingToken(ref StringSlice text)
+        {
+            ConsumeWhitespace(ref text);
+
+            foreach (Literal<ExpressionToken> token in Tokens)
+            {
+                if (text.StartsWith(token.Text, StringComparison.OrdinalIgnoreCase)) { return token; }
+            }
+
+            return null;
+        }
+
+        private static void ConsumeWhitespace(ref StringSlice text)
+        {
+            int whitespaceCount = 0;
+            for (; whitespaceCount < text.Length; ++whitespaceCount)
+            {
+                if (!IsWhitespace(text[whitespaceCount])) { break; }
+            }
+
+            text = text.Substring(whitespaceCount);
+        }
+
+        private static bool IsWhitespace(char c)
+        {
+            return c == ' ';
```

Comment:
>' ' [](start = 24, length = 3)

Only space?

Reply:
Yeah. I wanted to stick to just space as whitespace for now (JSON allows space, tab, carriage return, newline) but factored a method out in case we start accepting multiline expressions and wanted to expand it later.